builder/replica_pipeline_test.cc

```diff
@@ -0,0 +1,1129 @@
+  auto ab = buffer_expr::make(ctx, "ab", sizeof(int), 2);
+
+  var i(ctx, "i");
+  var j(ctx, "j");
+
+  // The bounds required of the dimensions consumed by the reduction depend on the size of the
+  // buffers passed in. Note that we haven't used any constants yet.
+  auto K_ab = a->dim(1).bounds;
+  auto K_abc = c->dim(0).bounds;
+
+  // We use int for this pipeline so we can test for correctness exactly.
+  func matmul_ab = func::make(matmul<int>, {{a, {point(i), K_ab}}, {b, {K_ab, point(j)}}}, {{ab, {i, j}}});
+  func matmul_abc = func::make(matmul<int>, {{ab, {point(i), K_abc}}, {c, {K_abc, point(j)}}}, {{abc, {i, j}}});
+
+  a->dim(1).stride = a->elem_size();
+  b->dim(1).stride = b->elem_size();
+  c->dim(1).stride = c->elem_size();
+  abc->dim(1).stride = abc->elem_size();
+
+  // TODO: There should be a more user friendly way to control the strides.
+  ab->dim(1).stride = static_cast<index_t>(sizeof(int));
+  ab->dim(0).stride = ab->dim(1).extent() * ab->dim(1).stride;
+
+  if (split > 0) {
+    matmul_abc.loops({{i, split, lm}});
+
+    if (lm == loop_mode::parallel) {
+      ab->store_at({&matmul_abc, i});
+    }
+  }
+
+  pipeline p = build_pipeline(ctx, {a, b, c}, {abc});
+  pipeline p_replica = kMatmulReplica();
+
+  // Look at the source code to this test to verify that we
+  // we have something that matches exactly
+  std::string replica_text = define_replica_pipeline(ctx, {a, b, c}, {abc});
+  LOG_REPLICA_TEXT(replica_text);
+  size_t pos = replica_pipeline_test_src.find(replica_text);
+  ASSERT_NE(pos, std::string::npos) << "Matching replica text not found, expected:\n" << replica_text;
+
+  // Run the pipeline
+  const int M = 10;
+  const int N = 10;
+  buffer<int, 2> a_buf({N, M});
+  buffer<int, 2> b_buf({N, M});
+  buffer<int, 2> c_buf({N, M});
+  buffer<int, 2> abc_buf({N, M});
+  // TODO: There should be a more user friendly way to initialize a buffer with strides other than the default
+  // order.
+  std::swap(a_buf.dim(1), a_buf.dim(0));
+  std::swap(b_buf.dim(1), b_buf.dim(0));
+  std::swap(c_buf.dim(1), c_buf.dim(0));
+  std::swap(abc_buf.dim(1), abc_buf.dim(0));
+
+  init_random(a_buf);
+  init_random(b_buf);
+  init_random(c_buf);
+  abc_buf.allocate();
+
+  const raw_buffer* inputs[] = {&a_buf, &b_buf, &c_buf};
+  const raw_buffer* outputs[] = {&abc_buf};
+  {
+    test_context eval_ctx;
+    p.evaluate(inputs, outputs, eval_ctx);
+  }
+  {
+    test_context eval_ctx;
+    p_replica.evaluate(inputs, outputs, eval_ctx);
+  }
+}
+
+// clang-format off
+static std::function<pipeline()> kPyramidReplica =
+// BEGIN define_replica_pipeline() output
+[]() -> ::slinky::pipeline {
+  node_context ctx;
+  auto in = buffer_expr::make(ctx, "in", sizeof(uint32_t), 2);
+  auto out = buffer_expr::make(ctx, "out", sizeof(uint32_t), 2);
+  auto x = var(ctx, "x");
+  auto y = var(ctx, "y");
+  auto intm = buffer_expr::make(ctx, "intm", sizeof(uint32_t), 2);
+  auto _2 = x * 2;
```

Comment:
I think printing expressions like this for replica pipelines is really verbose. I started on the replica pipeline before writing an issue for it, and I actually literally just used the existing print.h helper to print expressions. Can we do that?

Reply:
The existing print.h isn't designed to produce valid C++ code, and so things like undefined exprs and vars get emitted in a useless way. If we want to revise print.h to guarantee this I'll substitute it in, but we would need to commit to it. The expansion of verbosity here isn't ideal but for our purposes I don't think it really matters (this code isn't meant to be read in the same way as human-written code). If you want print.h revised in this way, I'll do so.

Reply:
You're right that print would generate incorrect syntax for unknown variables. I was thinking that bounds expressions do not reference variables other than the dims of the output (so you wouldn't get unknown variables), and that is very often true, but not always.

I guess we should just leave this for now, maybe we should modify print.h to generate more C compatible code.

The thing I'm thinking though is I really think this does matter, I expect we will be reading this generated code a lot. The way I expect to use this is that we will use this to generate standalone reproducers for bugs that occur in big complicated real world use cases. Debugging those bugs means reading this generated code (probably very carefully and repeatedly depending on the bug).

Reply:
I revised the code to inline more aggressively and use fewer explicit assignments. PTAL.